### ISSUE TITLE
Some improvements from another branch

### ISSUE THREAD

**Issue by racerxdl** [librtlsdr/librtlsdr#17]
@keenerd made some interesting changes in rtl_power and rtl_fm. We should take a look carefully at his changes and maybe merge to our branch. 

https://github.com/keenerd/rtl-sdr/

**Comment by hayguen** [comment]
the patches i saw look very interesting.

**Comment by ddcc** [comment]
Yes, it looks like he made a number of fixes and improvements, especially to `rtl_power`. I've been trying to track down a NaN output bug on ARM 32-bit, and it looks like it was fixed a while back in his branch: https://github.com/keenerd/rtl-sdr/commit/1b1c03d4e9330c30c9f4cc6de11fbe1572664fe7

**Comment by racerxdl** [comment]
@ddcc this changes makes a lot of sense. Since long size is not standard across platforms int64_t would be a best alternative. This weekend I will take a look in his changes and make a list of stuff that we need to port. Then anyone can pick a task to port something.

**Comment by Romeo-Golf** [comment]
+1

**Comment by pumatrax** [comment]
Yes please he has the sample mod in rtl_sdr

**Comment by hayguen** [comment]
cherry picked that commit .. and fixed merge conflicts  into the development branch.
please have a look/try!

**Comment by hayguen** [comment]
> Did you see this one keenerd/rtl-sdr@ebb5f2a ? If it is a bug, it is not yet fixed here and upstream.

don't think, that this was a real bug. took it anyway.

> These can be cherry picked keenerd/rtl-sdr@3dfa4ad, keenerd/rtl-sdr@d59b178.

applied rtl_fm: esbensen's discriminant, which looks good.
lib: cache i2c repeater: the 2nd patch was a bit more difficult:
* adapted for thread mutex .. thus, we can't do early exit
* removed all the additional rtlsdr_set_i2c_repeater() calls!
* checked all calls to rtlsdr_set_i2c_repeater()

other patches to follow ..

**Comment by hayguen** [comment]
> This keenerd/rtl-sdr@15bb16b fixed this keenerd/rtl-sdr@34ab4ac

the first 2 patches are for "rtl_fm: no-copy architecture". however it's not quickly done .. i get lockup/freeze, despite the first patch.
thus, i abandoned this .. one could give this another try ..

> this also needs to be applied keenerd/rtl-sdr@588b673 and this keenerd/rtl-sdr@42d8f12 and this keenerd/rtl-sdr@df2a9c0 and keenerd/rtl-sdr@88244c0

these ones are for "software agc (only for AM)":
i abandoned cherry-picking these also .. it looks strange / erroneous, cause it interferes with squelch:
* full_demod() does decimate/filter first, followed by the new software_agc() and (power)squelch. after those, the mode specific demodulation takes place. applying software_agc() before / alternative to squelch will break squelch function!
* IMHO software_agc() should take place after the mode specific demodulation .. that should also work for USB/LSB modes. but probably the comment in the usage() "only for AM" is simply wrong.

**Comment by hayguen** [comment]
> This activates stereo fm keenerd/rtl-sdr@53d212d

the commit title is "rtl_fm: half finished dual-channel mode",
and in the usage() you find "lrmix:  one channel goes to left audio, one to right (broken)".
thus, i did not pick this single commit.

for a full broadcast/wide FM stereo decoder, have a look at https://github.com/f4exb/ngsoftfm or better https://github.com/jj1bdx/ngsoftfm-jj1bdx , which also got replaced by https://github.com/jj1bdx/airspy-fmradion
last repo look active :-)

**Comment by hayguen** [comment]
> We should either apply keenerd/rtl-sdr@d426cfd

took that .. and improved it a bit: https://github.com/librtlsdr/librtlsdr/commit/27628aa4e20760b7c76ebf169247f1a1fa33b519

**Comment by ValZapod** [comment]
https://github.com/keenerd/rtl-sdr/commit/a2d23c590ff97821127d04c18009fab16f964d6f
I adpated patch for our case

```
diff --git a/src/rtl_fm.c b/src/rtl_fm.c
index abd7977..c092e5d 100644
--- a/src/rtl_fm.c
+++ b/src/rtl_fm.c
@@ -1316,9 +1316,9 @@ void full_demod(struct demod_state *d)
 
 static void rtlsdr_callback(unsigned char *buf, uint32_t len, void *ctx)
 {
-       struct dongle_state *s = ctx;
-       struct demod_state *d = s->demod_target;
-       struct cmd_state *c = d->cmd;
+       struct dongle_state *s;
+       struct demod_state *d;
+       struct cmd_state *c;
        int i, muteLen = s->mute;
        unsigned char sampleMax;
        uint32_t sampleP, samplePowSum = 0.0;
@@ -1329,6 +1329,9 @@ static void rtlsdr_callback(unsigned char *buf, uint32_t len, void *ctx)
                return;}
        if (!ctx) {
                return;}
+       s = ctx;
+       d = s->demod_target;
+       c = d->cmd;
        time(&rawtime);
        if (duration > 0 && rawtime >= stop_time) {
                do_exit = 1;


```

**Comment by ValZapod** [comment]
Wait a second, there is PPM calibration value in eeprom? Really??? https://github.com/keenerd/rtl-sdr/commit/389de6d8ed7e6c34e570bbac04c6007b4a19cc34

**Comment by hayguen** [comment]
> keenerd/rtl-sdr@a2d23c5    I adapted patch for our case

i prefer NOT to use this. this would mean libusb / librtlsdr is really broken and erroneous in passing the context pointer.
if this fails .. it will fail forever .. and IMHO it's better to see this problem with a crash!
despite that, i like the c++ style of directly initialize variables ..

**Comment by hayguen** [comment]
> Wait a second, there is PPM calibration value in eeprom? Really??? keenerd/rtl-sdr@389de6d

sounds cool for cheap dongles without TCXO :-)
i would prefer putting evaluation of serial number (for ppm value) and call to rtlsdr_set_freq_correction() directly into rtlsdr_open().
this way, one would not need to modify all the tools ..

thoughts on this?

**Comment by ValZapod** [comment]
> if this fails .. it will fail forever .. and IMHO it's better to see this problem with a crash!

It will **not** fail. Technically -> is allowed to dereference null pointer as it is **ACCESSING** the data on null pointer that is illegal, here there is no access. It will use offsetof macro instead in some cases in others there will be just no effect. It is a very crazy topic, and better to just not use it. https://stackoverflow.com/questions/28482809/c-access-static-members-using-null-pointer/28483477 There is also issue in standard http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#232

**Comment by ValZapod** [comment]
> this way, one would not need to modify all the tools ..

Is that a problem? You do not need all that reading in rtlsdr_open().

**Comment by hayguen** [comment]
> Is that a problem? You do not need all that reading in rtlsdr_open().

sorry, i don't understand your answer. can you explain?
what i meant, is not just my laziness in modification of tools, but you should also think of 3rd party tools, e.g. kal

**Comment by ValZapod** [comment]
> 4rd party tools, e.g. kal

Will not it broke them then? Also non-GPL programs need to use rtl_tcp.c, right?

**Comment by hayguen** [comment]
> Also non-GPL programs need to use rtl_tcp.c, right?

yes. another way might be using the piped output (of rtl_sdr or rtl_fm) ..


> Will not it broke them then?

possibly just in cases, where these already have some other mechanism to apply frequency correction.
but if these programs have a way to do so .. then these should also have a way to reset that.

in addition the user needs to change eeprom's serial number that 'automatic' frequency correction with rtlsdr_open() can take effect. the code is looking for serial to and with 'p' and looking for previous space .. and taking all that between as ppm value. from my experience, there is no factory serial ending with 'p' .. in contrary: all are the same '00000001' ! and i've tested several dongles (>10) with different tuners (FC0013, FC0012, E4000, R820T and R820T2).

furthermore, calls to rtlsdr_set_freq_correction() will still do as say should ..

**Comment by ValZapod** [comment]
> 'automatic' frequency correction with

Oh. So it is not really automatic? Is not done by hardware, i.e.? And also this is just a private format to store ppm value in serial number? Okay. Maybe it is not that interesting then.

**Comment by hayguen** [comment]
from what i see in the sources of the patch:

> So it is not really automatic?

not full automatic

> Is not done by hardware, i.e.?

just in sense of reading/writing to EEPROM.
but you still need to:
* run rtl_test, note result
* run rtl_eeprom to modify serial number with correctly formatted string containing the ppm value, e.g. "A 23p", for +23ppm A being serial "_number_". note: named serial "_number_" .. you might use non-digit characters

> And also this is just a private format to store ppm value in serial number?

yes, sort of 'misuse' of serial number:
have a serial code/string followed by a space, the ppm value and 'p'.

**Comment by ValZapod** [comment]
> sort of 'misuse' of serial number:

Okay, I do not like it, especially considering this: https://github.com/steve-m/librtlsdr/issues/33

So, another topic, is this possible bug? https://github.com/steve-m/librtlsdr/issues/12

**Comment by hayguen** [comment]
> especially considering this: steve-m#33

think, that in the end, it leads to this accepted PR: https://github.com/librtlsdr/librtlsdr/pull/66
it would be nice to have a good/automatic test procedure to create a plot as in for https://github.com/steve-m/librtlsdr/pull/10
there must be a test/detection program for detecting the frequency of this GSM 954.0 MHz FCCH burst!? does kal use that?

> is this possible bug? steve-m#12

the referenced hint https://github.com/steve-m/librtlsdr/commit/9e23e728c30e19c0d1e4e0b32f199679be9c2ac1 is: downcasting an index to **device_count** . for a device count, even an 8 bit counter would be sufficient .. this can't be relevant

**Comment by hayguen** [comment]
this thread is getting more than unreadable.
i would suggest creating new specific issues .. which might reference this one!

**Comment by ValZapod** [comment]
I think it is all. All other changes, like register cachinb and rtl_power are too compicated.

**Comment by hayguen** [comment]
thank you very much.

let me repeat the following - for new readers

> this thread is getting more than unreadable.
> i would suggest creating new specific issues .. which might reference this one!
